pkg/server/plugin/upstreamauthority/vault/vault.go

```diff
@@ -168,15 +167,23 @@ func (p *Plugin) MintX509CAAndSubscribe(req *upstreamauthorityv1.MintX509CAReque
 		return status.Error(codes.FailedPrecondition, "plugin not configured")
 	}
 
-	// reuseToken=false means that the token cannot be renewed and may expire,
-	// authenticates to the Vault at each signing request.
-	if p.vc == nil || !p.reuseToken {
-		vc, reusable, err := p.cc.NewAuthenticatedClient(p.authMethod)
+	vc := p.getVaultClient()
+
+	renewCh := make(chan struct{})
+	if vc == nil {
+		vc, err := p.cc.NewAuthenticatedClient(p.authMethod, renewCh)
 		if err != nil {
 			return status.Errorf(codes.Internal, "failed to prepare authenticated client: %v", err)
 		}
-		p.vc = vc
-		p.reuseToken = reusable
+		p.setVaultClient(vc)
+
+		// if renewCh has been closed, the token can not be renewed and may expire,
+		// it needs to re-authenticates to the Vault.
+		go func() {
+			<-renewCh
+			p.logger.Debug("Going to re-authenticate to the Vault at the next signing request time")
+			p.unsetVaultClient()
```

Comment:
Looking at the execution flow, I see in line 201:
`signResp, err := p.vc.SignIntermediate(ttl, csr)`
That's accessing `vc` outside of the mutex. It also concerns me that this is executed assuming that `vc` is not nil (there is an `if` that checks for that before) but there is a go routine that can set `vc` to nil before this line is executed, which would cause a panic.

Reply:
Your concern is correct. Thank to catch it.
3d2ff0b
I try to fix this.